### DISCUSSION TITLE
Pass a variable from Anylogic to Python

### DISCUSSION THREAD

**Discussion by MarloesAlb** [the-anylogic-company/AnyLogic-Pypeline#37] [Q&A] [answered]
I would like to run my Python file called GAresults with some inputs obtained from anylogic,
I have a variable in anylogic called varOrderIDList. This is a list, containg order id's. Like [1,2,3,4,5}
Now, my python file should use this list as input which is used in the script.
I tried 
pyCommunicator.run("import GAresults");
pyCommunicator.run("GAresults.order = %s", varOrderIDList, "GAresults.proc_time = %s", varOrderProcTime, "GAresults.due_dates = %s", varOrderDueDate);

But it just runs the file without using the assigned variables. In the python file, I defined order = [] proc_time=[] and due_dates=[]

Since if I do not assign a value in the python file, I get the error root.pyCommunicator:
NameError("name 'order' is not defined")

Which makes sense, as my python file first needs to now what ' order'  is (this is decided in anylogic, and not in python, so it is not defined in the python file. 

So, if the value of a variable is decided in anylogic, how could I pass that value to the python file?

I also tried

pyCommunicator.run("import GAresults", "GAresults.order = %s", [1,2,3,4], "GAresults.proc_time = %s", [5,6,4,6], "GAresults.due_dates = %s", [5,10,15,21]);

to see if that works,instead of using the variables. But I get the error Syntax error on token "[" , delete this token. Same for "]", so I also cannot pass the list of orderID's this way

**Comment by MarloesAlb** [comment, marked as answer]
I already found a solution! But I am not sure how to delete this question

**Reply by t-wolfeadam** [reply]
Glad to hear it! As I'm sure you found out, and in case anyone else has a similar problem, there were two issues here

1. The "syntax error" is because everything you pass to Python from Java needs to be in string form
2. There's a mix of the variants of `run` (i.e., you can either use multi-command _or_ the string replacement variant)

The easiest way to accomplish the above code is just the multi-command variant:

`pyCommunicator.run("import GAresults", "GAresults.order = [1,2,3,4]", "GAresults.proc_time = [5,6,4,6]", "GAresults.due_dates = [5,10,15,21]");`

But if you didn't know the list values, it's probably best to have a custom function in Python that sets the GAresults attributes based on three inputs. Then you can just call the single function + the `toJson` function for the arguments.

**Comment by AtikurIE** [comment]
Actually, I have studied your demo model and run it on my Python environment using TensorFlow as tf. Your demo model gives results for LOS in my desktop and Python environment imitating your demo model and data, I have trained a python model and tried to pass with AnyLogic. I have done almost and communicated my python using pipeline pycommunicator. It passes data from source to Delay block but after going delay block there is an error (root.pyCom:
_**Failed to run Python code; feedback: TypeError("'str' object is not callable")).**_ If this problem is solved I can use this model in my research. I have tried a lot to solve this problem but cannot cope up. I am sharing my zipfile folder attached in the email. Can you please help,check and run the model .

**Comment by AtikurIE** [comment]
[Regana.alp.zip](https://github.com/t-wolfeadam/AnyLogic-Pypeline/files/12286053/Regana.alp.zip)

**Comment by AtikurIE** [comment]
### Delay Block code:
pyCom.runResults(
double.class,
	 "predictor.RandomForestRegressor("   +  agent.getArrayStr()   +")" 
	)


### Function body of getArrayStr:

String placeholders = Strings.repeat("%s, ", 21).replaceFirst(", $", "");

return String.format("[["+placeholders+"]]", 
	
	numberOfIssues,
	(dialysisrenalendstage ? 1 : 0),
	(asthma ? 1 : 0),
	(irondef ? 1 : 0),
	(pneum ? 1 : 0),
	(substancedependence ? 1 : 0),
	(psychologicaldisordermajor ? 1 : 0),
	(depress ? 1 : 0),
	(psychother ? 1 : 0),
	(fibrosisandother ? 1 : 0),
	(malnutrition ? 1 : 0),
	(hemo ? 1 : 0),
	hematocritN ,
	 neutrophilsN  ,
	 sodiumN  ,
	 glucoseN ,
	 bloodureanitroN  ,
	creatinineN  ,
	bmiN ,
	 pulseN ,
	respirationN ,
	secondarydiagnosisnonicd9 
	);	
**Function body of ToStr:**
return String.format("patient %3d %s", getId(), getArrayStr());

**code at py file:**


import numpy as np
from sklearn.ensemble import RandomForestRegressor
from collections import deque

class classifier:
    ''' A simple class to easily query the neural networks from inside AnyLogic using Pypeline '''
    def __init__(self):
        # load the LOS model
        self.RandomForestRegressor = ("trained_random_forest_model.h5")
       
    def RandomForestRegressor (self, patient_data):
        ''' Given the (1, 21) array of patient data, predict the length of stay (days) '''
        # convert default list to numpy array
        patient_array = np.array(patient_data)
        
        # query the neural network for the length of stay
        prediction = self.trained_random_forest_model.predict(patient_array)
        return prediction[0][0]
        
**please give me a solution why bethe error coming:**
Failed to run Python code; feedback: TypeError("'str' object is not callable")).
